contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/io/KafkaSource.java

```diff
@@ -0,0 +1,251 @@
+   * require deduping, use withDedup() when constructing the KafkaSource.
+   */
+  @Override
+  public boolean requiresDeduping() {
+    return dedup;
+  }
+
+
+  /**
+   * A {@code Reader} that reads unbounded input from a single Kafka topic.
+   */
+  private class KafkaSourceReader extends UnboundedReader<ConsumerRecord<String, String>> {
+
+    private final KafkaCheckpoint checkpoint;
+
+    KafkaSourceReader(KafkaCheckpoint checkpoint,
+        String bootstrapServers, String groupId, String topic) {
+      if (checkpoint == null) {
+        this.checkpoint = new KafkaCheckpoint(bootstrapServers, groupId, topic);
+      } else {
+        this.checkpoint = checkpoint;
+      }
+    }
+
+    /**
+     * Initializes the reader and advances the reader to the first record.
+     */
+    @Override
+    public boolean start() {
+      return this.advance();
+    }
+
+    /**
+     * Advances the reader to the next valid record.
+     *
+     * <p>Returns {@code true} if a record was read, {@code false} if there is no more input
+     * available. Future calls to {@link #advance} may return {@code true} once more data is
+     * available.
+     */
+    @Override
+    public boolean advance() {
+      return this.checkpoint.advance();
+    }
+
+    /**
+     * Returns the data record at the current position, last read by calling start or advance.
+     */
+    @Override
+    public ConsumerRecord<String, String> getCurrent() throws NoSuchElementException {
+      return this.checkpoint.current();
+    }
+
+    /**
+     * Returns a unique identifier for the current record based
+     * on the current offset, topic and partition.
+     */
+    @Override
+    public byte[] getCurrentRecordId() throws NoSuchElementException {
+      try {
+        ConsumerRecord<String, String> record = getCurrent();
+        if (record != null) {
+          return CoderUtils.encodeToByteArray(StringUtf8Coder.of(), record.toString());
+        } else {
+          return new byte[0];
+        }
+      } catch (CoderException e) {
+        LOG.error(Throwables.getStackTraceAsString(e));
+        return new byte[0];
+      }
+    }
+
+    /**
+     * Returns a lower bound on timestamps of future elements read by this
+     * reader.
+     *
+     * Kafka records do not have a natural timestamp so use clock time instead.
+     */
+    @Override
+    public Instant getWatermark() {
+      DateTime dt = new DateTime(DateTimeZone.UTC);
+      return dt.toInstant();
+    }
+
+    /**
+     * Returns the timestamp associated with the current data item.
+     *
+     * Kafka records do not have a natural timestamp so min value instead.
+     */
+    @Override
+    public Instant getCurrentTimestamp() throws NoSuchElementException {
+      ConsumerRecord<String, String> record = getCurrent();
+      if (record == null) {
+        throw new NoSuchElementException();
+      }
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
```

Comment:
If you always return the `TIMESTAMP_MIN_VALUE` then I think your streaming pipeline will stall in practice.

A usual strategy is to let users specify a `SerializableFunction<T, Instant>` to pull the timestamp out of the element if there is one in the payload, or to use `Instant.now()` when the element is pulled from Kafka. (Note that successive calls to `getCurrentTimestamp()` should return the same value, so using `Instant.now()` here would not be semantically correct.